File: src/tiandao/audit_event.py
"""Neutral Tiandao AuditEvent contract merged into the Python mirror."""

from __future__ import annotations

import random
import time
from datetime import datetime, timezone
from enum import Enum
from typing import Any

UTC = timezone.utc


class AuditAction(str, Enum):
    TASK_CREATE = "task.create"
    TASK_EXECUTE = "task.execute"
    TASK_COMPLETE = "task.complete"
    TASK_FAIL = "task.fail"
    VERDICT_CREATE = "verdict.create"
    VERDICT_APPROVE = "verdict.approve"
    VERDICT_REJECT = "verdict.reject"
    ROLE_CREATE = "role.create"
    ROLE_ASSIGN = "role.assign"
    ROLE_REVOKE = "role.revoke"
    CONTEXT_ASSEMBLE = "context.assemble"
    CONTEXT_INJECT = "context.inject"
    EVIDENCE_COLLECT = "evidence.collect"
    SOURCE_REF_TRACK = "source_ref.track"
    CONNECTION_OPEN = "connection.open"
    CONNECTION_CLOSE = "connection.close"
    REGISTRY_REGISTER = "registry.register"
    REGISTRY_UNREGISTER = "registry.unregister"


class AuditResult(str, Enum):
    SUCCESS = "success"
    FAILURE = "failure"
    PENDING = "pending"
    PARTIAL = "partial"


def _audit_ts() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def _audit_event_id() -> str:
    suffix = "".join(random.choice("abcdefghijklmnopqrstuvwxyz0123456789") for _ in range(8))
    return f"audit_{int(time.time() * 1000)}_{suffix}"


class TiandaoAuditEvent:
    """Common audit envelope shared by Honghuang systems."""

    def __init__(
        self,
        actor: str,
        action: AuditAction | str,
        target: str,
        result: AuditResult | str,
        event_id: str = "",
        timestamp: str = "",
        details: dict[str, Any] | None = None,
        source_ip: str = "",
        user_agent: str = "",
    ):
        self.event_id = event_id or _audit_event_id()
        self.timestamp = timestamp or _audit_ts()
        self.actor = actor
        self.action = AuditAction(action)
        self.target = target
        self.result = AuditResult(result)
        self.details = details or {}
        self.source_ip = source_ip
        self.user_agent = user_agent

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "event_id": self.event_id,
            "timestamp": self.timestamp,
            "actor": self.actor,
            "action": self.action.value,
            "target": self.target,
            "result": self.result.value,
        }
        if self.details:
            data["details"] = self.details
        if self.source_ip:
            data["source_ip"] = self.source_ip
        if self.user_agent:
            data["user_agent"] = self.user_agent
        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "TiandaoAuditEvent":
        return cls(
            event_id=str(data.get("event_id", "")),
            timestamp=str(data.get("timestamp", "")),
            actor=str(data.get("actor", "")),
            action=str(data.get("action", AuditAction.TASK_CREATE.value)),
            target=str(data.get("target", "")),
            result=str(data.get("result", AuditResult.PENDING.value)),
            details=data.get("details") if isinstance(data.get("details"), dict) else {},
            source_ip=str(data.get("source_ip", "")),
            user_agent=str(data.get("user_agent", "")),
        )


def create_audit_event(
    *,
    actor: str,
    action: AuditAction | str,
    target: str,
    result: AuditResult | str,
    event_id: str = "",
    details: dict[str, Any] | None = None,
    source_ip: str = "",
    user_agent: str = "",
) -> dict[str, Any]:
    return TiandaoAuditEvent(
        event_id=event_id,
        actor=actor,
        action=action,
        target=target,
        result=result,
        details=details,
        source_ip=source_ip,
        user_agent=user_agent,
    ).to_dict()

File: src/tiandao/contracts/audit-event.ts
/**
 * 公共 AuditEvent — 共用合同
 * Tiandao AuditEvent Schema
 *
 * 设计口径：
 *   公共规则是跨系统的中性公共规则层，随着各系统实践而完善。
 *   各系统共用一套公共规则合同，只使用自己需要的公共规则面。
 *
 *   AuditEvent 是跨系统通用的审计事件信封。
 *   不归属于任何单个子系统。
 *   归属公共规则层。
 *
 *   旧系统内的同名类型只应作为兼容 re-export，不拥有公共合同。
 */

/** 审计动作类型 */
export type AuditAction =
  | 'task.create'
  | 'task.execute'
  | 'task.complete'
  | 'task.fail'
  | 'verdict.create'
  | 'verdict.approve'
  | 'verdict.reject'
  | 'role.create'
  | 'role.assign'
  | 'role.revoke'
  | 'context.assemble'
  | 'context.inject'
  | 'evidence.collect'
  | 'source_ref.track'
  | 'connection.open'
  | 'connection.close'
  | 'registry.register'
  | 'registry.unregister';

/** 审计事件结果 */
export type AuditResult = 'success' | 'failure' | 'pending' | 'partial';

/** AuditEvent — 公共规则共用审计事件 */
export interface TiandaoAuditEvent {
  /** 事件 ID */
  event_id: string;
  /** 时间戳 */
  timestamp: string; // ISO datetime
  /** 执行者 */
  actor: string;
  /** 动作类型 */
  action: AuditAction;
  /** 目标对象 */
  target: string;
  /** 执行结果 */
  result: AuditResult;
  /** 详情 */
  details?: Record<string, unknown>;
  /** 来源 IP */
  source_ip?: string;
  /** User Agent */
  user_agent?: string;
}

/** 创建审计事件的工厂函数 */
export function createAuditEvent(params: {
  eventId?: string;
  actor: string;
  action: AuditAction;
  target: string;
  result: AuditResult;
  details?: Record<string, unknown>;
  sourceIp?: string;
  userAgent?: string;
}): TiandaoAuditEvent {
  return {
    event_id: params.eventId || `audit_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
    timestamp: new Date().toISOString(),
    actor: params.actor,
    action: params.action,
    target: params.target,
    result: params.result,
    details: params.details,
    source_ip: params.sourceIp,
    user_agent: params.userAgent,
  };
}
